File: 3455.shortest-matching-substring.java
#
# @lc app=leetcode id=3455 lang=java
#
# [3455] Shortest Matching Substring
#

# @lc code=start
class Solution {
    public int shortestMatchingSubstring(String s, String p) {
        int n = s.length();
        int star1 = p.indexOf('*');
        int star2 = p.indexOf('*', star1 + 1);
        String A = p.substring(0, star1);
        String B = p.substring(star1 + 1, star2);
        String C = p.substring(star2 + 1);
        int la = A.length();
        int lb = B.length();
        int lc = C.length();
        java.util.ArrayList<Integer> posA = kmp(s, A);
        java.util.ArrayList<Integer> posB = kmp(s, B);
        java.util.ArrayList<Integer> posC = kmp(s, C);
        int minLen = Integer.MAX_VALUE;
        for (int sb : posB) {
            int maxl = sb - la;
            if (maxl < 0) continue;
            int l = findLargestLE(posA, maxl);
            if (l == -1) continue;
            int minsc = sb + lb;
            int sc = findSmallestGE(posC, minsc);
            if (sc == -1) continue;
            int r = sc + lc - 1;
            int curlen = r - l + 1;
            if (curlen >= 0 && curlen < minLen) {
                minLen = curlen;
            }
        }
        return minLen == Integer.MAX_VALUE ? -1 : minLen;
    }

    private java.util.ArrayList<Integer> kmp(String text, String pat) {
        int n = text.length();
        int m = pat.length();
        java.util.ArrayList<Integer> res = new java.util.ArrayList<>();
        if (m == 0) {
            for (int i = 0; i <= n; i++) {
                res.add(i);
            }
            return res;
        }
        int[] pi = computePrefixFunction(pat);
        int q = 0;
        for (int i = 0; i < n; i++) {
            while (q > 0 && pat.charAt(q) != text.charAt(i)) {
                q = pi[q - 1];
            }
            if (pat.charAt(q) == text.charAt(i)) {
                ++q;
            }
            if (q == m) {
                res.add(i - m + 1);
                q = pi[q - 1];
            }
        }
        return res;
    }

    private int[] computePrefixFunction(String pat) {
        int m = pat.length();
        int[] pi = new int[m];
        int k = 0;
        for (int i = 1; i < m; i++) {
            while (k > 0 && pat.charAt(k) != pat.charAt(i)) {
                k = pi[k - 1];
            }
            if (pat.charAt(k) == pat.charAt(i)) {
                ++k;
            }
            pi[i] = k;
        }
        return pi;
    }

    private int findLargestLE(java.util.ArrayList<Integer> list, int target) {
        int lo = 0;
        int hi = list.size() - 1;
        int ansIdx = -1;
        while (lo <= hi) {
            int mid = lo + (hi - lo) / 2;
            if (list.get(mid) <= target) {
                ansIdx = mid;
                lo = mid + 1;
            } else {
                hi = mid - 1;
            }
        }
        return ansIdx == -1 ? -1 : list.get(ansIdx);
    }

    private int findSmallestGE(java.util.ArrayList<Integer> list, int target) {
        int lo = 0;
        int hi = list.size() - 1;
        int ansIdx = -1;
        while (lo <= hi) {
            int mid = lo + (hi - lo) / 2;
            if (list.get(mid) >= target) {
                ansIdx = mid;
                hi = mid - 1;
            } else {
                lo = mid + 1;
            }
        }
        return ansIdx == -1 ? -1 : list.get(ansIdx);
    }
}
# @lc code=end

File: 3455.shortest-matching-substring.go
#
# @lc app=leetcode id=3455 lang=golang
#
# [3455] Shortest Matching Substring
#

# @lc code=start
func shortestMatchingSubstring(s string, p string) int {
	star1 := -1
	star2 := -1
	for i := 0; i < len(p); i++ {
		if p[i] == '*' {
			if star1 == -1 {
				star1 = i
			} else {
				star2 = i
				break
			}
		}
	}

	a := p[:star1]
	b := p[star1+1:star2]
	c := p[star2+1:]
	la, lb, lc := len(a), len(b), len(c)

	computeLPS := func(pat string) []int {
		m := len(pat)
		if m == 0 {
			return []int{}
		}
		pi := make([]int, m)
		j := 0
		i := 1
		for i < m {
			if pat[i] == pat[j] {
				j++
				pi[i] = j
				i++
			} else {
				if j != 0 {
					j = pi[j-1]
				} else {
					pi[i] = 0
					i++
				}
			}
		}
		return pi
	}

	findAllMatches := func(text, pattern string) []int {
		n := len(text)
		m := len(pattern)
		if m == 0 {
			res := make([]int, n+1)
			for i := 0; i <= n; i++ {
				res[i] = i
			}
			return res
		}
		pi := computeLPS(pattern)
		res := []int{}
		j := 0
		for i := 0; i < n; i++ {
			for j > 0 && text[i] != pattern[j] {
				j = pi[j-1]
			}
			if text[i] == pattern[j] {
				j++
			}
			if j == m {
				res = append(res, i-m+1)
				j = pi[j-1]
			}
		}
		return res
	}

	largestLE := func(pos []int, val int) int {
		lo := 0
		hi := len(pos) - 1
		res := -1
		for lo <= hi {
			mid := lo + (hi-lo)/2
			if pos[mid] <= val {
				res = mid
				lo = mid + 1
			} else {
				hi = mid - 1
			}
		}
		return res
	}

	smallestGE := func(pos []int, val int) int {
		lo := 0
		hi := len(pos) - 1
		res := -1
		for lo <= hi {
			mid := lo + (hi-lo)/2
			if pos[mid] >= val {
				res = mid
				hi = mid - 1
			} else {
				lo = mid + 1
			}
		}
		return res
	}

	posa := findAllMatches(s, a)
	posb := findAllMatches(s, b)
	posc := findAllMatches(s, c)

	inf := 1 << 30
	minlen := inf
	for j := 0; j < len(posb); j++ {
		startb := posb[j]
		limitl := startb - la
		idxl := largestLE(posa, limitl)
		if idxl == -1 {
			continue
		}
		startr := startb + lb
		idxr := smallestGE(posc, startr)
		if idxr == -1 {
			continue
		}
		cand := posc[idxr] + lc - posa[idxl]
		if cand < minlen {
			minlen = cand
		}
	}
	if minlen == inf {
		return -1
	}
	return minlen
}
# @lc code=end